Maple2.Server.Game/Manager/Config/BuffManager.cs

```diff
@@ -133,30 +152,65 @@ public void AddBuff(IActor caster, IActor owner, int id, short level, long start
         }
 
         logger.Information("{Id} AddBuff to {ObjectId}: {SkillId},{Level} for {Tick}ms", buff.ObjectId, owner.ObjectId, id, level, buff.EndTick - buff.StartTick);
-        // Logger.Information("> {Data}", additionalEffect.Property);
         if (owner is FieldPlayer player) {
+            if (additionalEffect.Property.Exp > 0) { // Assuming this doesnt proc s
+                player.Session.Exp.AddStaticExp(additionalEffect.Property.Exp);
+            }
+
             player.Session.ConditionUpdate(ConditionType.buff, codeLong: buff.Id);
             player.Session.Dungeon.UpdateDungeonEnterLimit();
         }
         if (notifyField) {
             owner.Field.Broadcast(BuffPacket.Add(buff));
         }
+        SetMount(buff);
     }
 
-    public bool HasBuff(int effectId, short effectLevel = 1, int overlapCount = 0) {
-        if (!Buffs.TryGetValue(effectId, out Buff? buff)) {
+    private Buff? GetBuff(int buffId, int casterObjectId = 0) {
+        if (casterObjectId > 0) {
+            if (Buffs.TryGetValue(buffId, out List<Buff>? buffs)) {
+                return buffs.FirstOrDefault(buff => buff.Caster.ObjectId == casterObjectId);
+            }
+        } else {
+            if (Buffs.TryGetValue(buffId, out List<Buff>? buffs)) {
+                return buffs.FirstOrDefault();
+            }
+        }
+        return null;
+    }
+
+    private bool TryAdd(Buff buff) {
+        if (Buffs.TryGetValue(buff.Id, out List<Buff>? buffs)) {
+            Buff? casterBuff = buffs.FirstOrDefault(b => b.Caster.ObjectId == buff.Caster.ObjectId);
+            if (casterBuff != null) {
+                return false;
+            }
+            buffs.Add(buff);
+            return true;
+        }
+
+        return Buffs.TryAdd(buff.Id, [buff]);
+    }
+
+    public List<Buff> EnumerateBuffs() => Buffs.Values.SelectMany(list => list).ToList();
+    public List<Buff> EnumerateBuffs(int buffId) => Buffs.TryGetValue(buffId, out List<Buff>? buffs) ? buffs : [];
```

Comment:
idk if there's any side effects here but maybe not making a list and leaving it as IEnumerable could help?

for example:
```cs
List<Buff> buffs = EnumerateBuffs(modifyOverlapCount.Id);
foreach (Buff buffResult in buffs) {
  if (buffResult.Stack(modifyOverlapCount.OffsetCount)) {
    Actor.Field.Broadcast(BuffPacket.Update(buffResult));
  }
}
```
could become

```cs
IEnumerable<Buff> buffs = EnumerateBuffs(modifyOverlapCount.Id).Where(buff => buff.Stack(modifyOverlapCount.OffsetCount));
foreach (Buff buffResult in buffs) {
    Actor.Field.Broadcast(BuffPacket.Update(buffResult));
}
```

this example happens in multiple places